sno/working_copy/sqlserver.py

```diff
@@ -0,0 +1,553 @@
+import contextlib
+import logging
+import re
+import time
+from urllib.parse import urlsplit, urlunsplit
+
+import click
+import sqlalchemy
+from sqlalchemy import literal_column
+from sqlalchemy.ext.compiler import compiles
+from sqlalchemy.orm import sessionmaker
+from sqlalchemy.sql import crud, quoted_name
+from sqlalchemy.sql.dml import ValuesBase
+from sqlalchemy.sql.functions import Function
+from sqlalchemy.sql.compiler import IdentifierPreparer
+from sqlalchemy.types import UserDefinedType
+
+from .base import WorkingCopy
+from . import sqlserver_adapter
+from .table_defs import SqlServerSnoTables
+from sno import crs_util
+from sno.geometry import Geometry
+from sno.exceptions import InvalidOperation
+from sno.sqlalchemy import sqlserver_engine
+
+
+"""
+* database needs to exist
+* database needs to have postgis enabled
+* database user needs to be able to:
+    1. create 'sno' schema & tables
+    2. create & alter tables in the default (or specified) schema
+    3. create triggers
+"""
+
+L = logging.getLogger("sno.working_copy.postgis")
+
+
+class WorkingCopy_SqlServer(WorkingCopy):
+    def __init__(self, repo, uri):
+        """
+        uri: connection string of the form mssql://[user[:password]@][netloc][:port][/dbname/schema][?param1=value1&...]
+        """
+        self.L = logging.getLogger(self.__class__.__qualname__)
+
+        self.repo = repo
+        self.uri = uri
+        self.path = uri
+
+        url = urlsplit(uri)
+
+        if url.scheme != "mssql":
+            raise ValueError("Expecting mssql://")
+
+        url_path = url.path
+        path_parts = url_path[1:].split("/", 3) if url_path else []
+        if len(path_parts) != 2:
+            raise ValueError("Expecting mssql://[HOST]/DBNAME/SCHEMA")
```

Comment:
URL standard:
This URL isn't any kind of formal standard - it seems to be a sqlalchemy thing - some DB's have sensible URLs, like postgres://, some don't, like SQL server, but sqlalchemy accepts sensible looking URLs even for SQL server and converts them to whatever crazy thing SQL server expects.

Us adding the schema onto the end is also not a standard, but I like it better than a query-string - logically, the next path segment after the database is the database-schema, and so on if there are smaller subcontainers inside that. Just my preference.

Requiring a schema:
I'll keep this as-is for this PR - I can change it in a separate PR if we need to, both for this and for PostGIS. I think I convinced Rob last year that we should take ownership of a schema, although the schema-parsing code is originally his. Giving Sno complete ownership of a db schema makes good sense conceptually and avoids some difficulties:

- git owns a worktree folder, sno owns a GPKG - or a db schema. These are well-defined containers that are wholly owned by a single repo, everything inside them is managed.
- It's clear which tables are owned by which sno repo, instead of them being in a big pile with every other table
- Sno needs to create and delete tables - this avoids collisions with user tables of the same name, etc.
- although its not yet supported, we might one day allow tables to be created by the user inside the working copy and committed directly, instead of being created elsewhere and then imported. Having a clear boundary between the inside and outside of the working copy will make this possible.
- Avoids any mixups where tables end up being managed by two repos at once